### PR TITLE
Add support for an installation mirror of Node.js binaries

### PR DESCRIPTION
**Pull request by trippingtarballs**
**Task name**: NodeTool

**Description**: Add support for the installation of Node.js via mirror.

**Documentation changes required:** Y

**Added unit tests:** N

**Attached related issue:** #17575

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

---

This PR is intended to parameterise the mirror used when sourcing the installation package for Node.js. This is achieved by setting `nodejsMirror`, a new optional input to NodeTool. The default value for this input is `https://nodejs.org/dist`. This hardcoded value was been replaced with a variable in this PR.

### CODE REVIEW

**Comment by trippingtarballs** [comment]
@microsoft-github-policy-service agree

**Review by KonstantinTyukalov — APPROVED** [review]

File: Tasks/NodeToolV0/nodetool.ts
Comment (ismayilov-ismayil) [inline Tasks/NodeToolV0/nodetool.ts:129]: If provided URL contains slash (/) at the end it will not work. Could you please trim slashes at the end before using it?

File: Tasks/NodeToolV0/task.json
Comment (ismayilov-ismayil) [inline Tasks/NodeToolV0/task.json:73]: My suggestion is to move this option to advanced group, and change the description. Because when we have default value 'an alternative source' seems a bit stranger.

Comment (trippingtarballs) [reply Tasks/NodeToolV0/task.json]: > "move this option to advanced group"

What does "advanced group" mean? I have searched the codebase and cannot find what "advanced group" might refer to.

I have reworded the text. If it is not to your satisfaction then please may I ask that you provide an alternative suggestion.

Comment (ismayilov-ismayil) [reply Tasks/NodeToolV0/task.json]: You can refer MSBuildV1 task.json, there might be what I prefer. 

Regarding label my suggestion is "Set source for Node.js binaries" and you can move current label to helpMarkdown to give more detailed information about this input

Comment (trippingtarballs) [reply Tasks/NodeToolV0/task.json]: > "... refer MSBuildV1 task.json,"

[(commit)](https://github.com/microsoft/azure-pipelines-tasks/pull/17500/commits/b46553a65388b4280a1366aeb61bc5f25b5240ea) I have created a new Advanced Group and marked the `nodejsMirror` input as belonging to that `"advanced"` group. 

> "my suggestion is 'Set source for Node.js binaries'"

[(commit)](https://github.com/microsoft/azure-pipelines-tasks/pull/17500/commits/35f00f952ddf1d8c3e8f406d3635bd37d194a3aa) I have reworded the value of the `label` text to exactly what you have asked.

> "can move current label to helpMarkdown"

Sorry @ismayilov-ismayil, I do not understand what you are asking me to do. Yes, I have a `label` value and I have a `helpMarkdown` for the new input: `nodejsMirror`, and I have set the `label` value to the text you wanted. Are you asking me to remove `label` completely and include only a `helpMarkdown` field with the text you have suggested? Please clarify.

Comment (ismayilov-ismayil) [reply Tasks/NodeToolV0/task.json]: @trippingtarballs sorry if I couldn't explain clearly. 
You are already done what I suggested. Thanks

File: Tasks/NodeToolV0/task.json
Comment (ismayilov-ismayil) [inline Tasks/NodeToolV0/task.json:16]: Please, bump version to 217, then we can continue. thanks

**Comment by trippingtarballs** [comment]
@ismayilov-ismayil, I believe all code-review comments have been addressed. I have [bumped](https://github.com/microsoft/azure-pipelines-tasks/pull/17500/commits/7ab41c52c31ce30b946bbd899c5c99b9e7bfccfd#diff-aada123c47bc7fec10b28b3b25e485fcdfe0800687619ce354d46a2793ed702aR16) the `version` as requested. I am happy to mark all comments as resolved. Thank you for your support with this PR -- I am excited to get it merged!

**Review by ismayilov-ismayil — APPROVED** [review]

**Comment by ismayilov-ismayil** [comment]
@trippingtarballs we have issue with .replace of nodeJsMirror, could you please run
```  node make.js build  --task NodeToolV0 ```
and 
```node make.js test --task NodeToolV0 ```
locally to see the error and fix it

**Comment by trippingtarballs** [comment]
@ismayilov-ismayil, sure. I'll do it now. However, before I do, may I ask you to confirm the version of Node.js and NPM I should have setup. 

It is unclear if the [docs](https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/contribute.md#prerequisites-node-and-npm) are out of date. They appear to say Node.js at v4 and NPM and v5, and then say Node.js at v10 🤷  

What version of Node.js and NPM are you using on your system for development on this project?

**Comment by ismayilov-ismayil** [comment]
> @ismayilov-ismayil, sure. I'll do it now. However, before I do, may I ask you to confirm the version of Node.js and NPM I should have setup.
> 
> It is unclear if the [docs](https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/contribute.md#prerequisites-node-and-npm) are out of date. They appear to say Node.js at v4 and NPM and v5, and then say Node.js at v10 🤷
> 
> What version of Node.js and NPM are you using on your system for development on this project?

Node 10 is fine

File: Tasks/NodeToolV0/Tests/L0FirstDownloadSuccess.ts
Comment (trippingtarballs) [inline Tasks/NodeToolV0/Tests/L0FirstDownloadSuccess.ts:12]: @ismayilov-ismayil, 

The fix I have put in place to make the tests pass is to add this line:

```ts
tmr.setInput('nodejsMirror', 'https://nodejs.org/dist');
```

It's the same line I've added to all of the test files for NodeTool. 

However, this fix only makes sense if the following assumption is true: **The execution of the tests for NodeTool do not load the [default value](https://github.com/microsoft/azure-pipelines-tasks/pull/17500/files#diff-aada123c47bc7fec10b28b3b25e485fcdfe0800687619ce354d46a2793ed702aR81) for the inputs**. I believe this assumption is correct because this is on the line above my added line:

```ts
tmr.setInput('checkLatest', 'false');
```

And from the `master` branch you can see that `checkLatest` has a [default value](https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/NodeToolV0/task.json#L58) of `"false"`. 

If the tests for NodeTool were being primed with the default values defined in the `task.json` then my fix wouldn't be needed (nor would `checkLatest` need to be set to `false` in the tests).

**Comment by trippingtarballs** [comment]
@ismayilov-ismayil It looks like the build is failing .. am I understanding this correctly? Do you need me to take any further actions here?

**Review by sergey-koryshev — APPROVED** [review]

**Review by kirill-ivlev — APPROVED** [review]

**Comment by ismayilov-ismayil** [comment]
Hi @trippingtarballs 
could you please bump the task version to 218? 
Sorry for inconvenience.

**Review by wojgasior — APPROVED** [review]

**Comment by trippingtarballs** [comment]
@ismayilov-ismayil I have bumped to `218`.

Why is PR lingering and not being merged? Am I doing something wrong that make the PR not ready to be merged? Do you need me to keep merging `master` into my feature branch?

**Comment by ismayilov-ismayil** [comment]
@trippingtarballs thank you for your contribution and support. We need to get all corresponding teams approval, thus sorry for delay. 
Your next PRs are welcome ;)

**Comment by trippingtarballs** [comment]
Wonderful, it's been merged! 🥳   When can I start using this feature @ismayilov-ismayil?

**Comment by trippingtarballs** [comment]
@Roman-Shchukin / @ismayilov-ismayil, please advise me how I can use `NodeTool@0.218.0`.  Do I need to wait for a release? When does NodeTool get released?

**Pull request merged**
